internal-dns/src/resolver.rs

```diff
@@ -515,4 +588,219 @@ mod test {
+
+        server.close().await.expect("Failed to stop test server");
+        dns_server.cleanup_successful();
+        logctx.cleanup_successful();
+    }
+
+    #[tokio::test]
+    async fn resolver_can_access_backup_dns_server() {
+        let logctx = test_setup_log("resolver_can_access_backup_dns_server");
+
+        // Confirm that we can create a progenitor client for this server.
+        expect_openapi_json_valid_for_test_server();
+
+        // Create DNS servers, and a corresponding resolver.
+        let dns_server1 = DnsServer::create(&logctx.log).await;
+        let dns_server2 = DnsServer::create(&logctx.log).await;
+        let resolver = Resolver::new_from_addrs(
+            logctx.log.clone(),
+            vec![
+                *dns_server1.dns_server.local_address(),
+                *dns_server2.dns_server.local_address(),
+            ],
+        )
+        .unwrap();
+
+        // Start a test server, but don't register it with the DNS server (yet).
+        let label = 1234;
+        let server = start_test_server(logctx.log.clone(), label);
+        let ip = match server.local_addr().ip() {
+            std::net::IpAddr::V6(ip) => ip,
+            _ => panic!("Expected IPv6"),
+        };
+        let port = server.local_addr().port();
+
+        // Use the resolver -- referencing our DNS server -- in the construction
+        // of a progenitor client.
+        //
+        // We'll use the SRV record for Nexus, even though it's just our
+        // standalone test server.
+        let dns_name = crate::ServiceName::Nexus.srv_name();
+        let reqwest_client = reqwest::ClientBuilder::new()
+            .dns_resolver(resolver.clone().into())
+            .build()
+            .expect("Failed to build client");
+
+        // NOTE: We explicitly pass the port here, before DNS resolution,
+        // because the DNS support in reqwest does not actually use the ports
+        // returned by the resolver.
+        let client = Client::new_with_client(
+            &format!("http://{dns_name}:{port}"),
+            reqwest_client,
+            logctx.log.clone(),
+        );
+
+        // The DNS server is running, but has no records. Expect a failure.
+        let err = client.test_endpoint().await.unwrap_err();
+        assert!(
+            err.to_string().contains("no record found"),
+            "Unexpected Error (expected 'no record found'): {err}",
+        );
+
+        // Add a record for the new service, but only to the second DNS server.
+        let mut dns_config = DnsConfigBuilder::new();
+        let zone = dns_config.host_zone(Uuid::new_v4(), ip).unwrap();
+        dns_config
+            .service_backend_zone(ServiceName::Nexus, &zone, port)
+            .unwrap();
+        let dns_config = dns_config.build();
+        dns_server2.update(&dns_config).await.unwrap();
+
+        // Confirm that the progenitor client can access this record,
+        // even though the first DNS server doesn't know about it.
```

Comment:
I'm not sure that this should necessarily work.  I think all DNS servers are equally authoritative.  If this didn't work, I don't think anything would be broken.  (A related test would be if you _shut down_ the first server -- then I think it should work.)

Reply:
I'll shut down the first server first -- the test still passes with that tweak.